# Программирование на Python 

# Основные типы данных. Создание переменных. Основные ошибки.

*Автор: Паршина Анастасия, НИУ ВШЭ (tg: @aaparshina)*

*Дополнила: Капустина Лика, НИУ ВШЭ (tg: @lika_kapustina)*

## Содержание
1. [Создание переменных и оператор присваивания](#par1)
2. [Изменяемые (mutable) и неизменяемые (immutable) типы данных](#par2)
3. [Типы данных — часть 1](#par3)
    1. [Числа](#par3.1)
    2. [Строки](#par3.2)
    3. [Преобразование типов](#par3.3)
4. [Немного об ошибках](#par4)
5. [Разметка Markdown в Jupyter Notebook](#par5)
6. [Дополнительные материалы](#parlast)

## Создание переменных и оператор присваивания <a name="par1"></a>

Данные в Python представляют собой объекты и отношения между объектами. У каждого объекта есть свой индентификатор (id), тип и значение.

<p style="text-align: center;"><b>имя переменной = объект</b></p>

Оператор присваивания (то есть наш знак `=`) запоминает за именем переменной идентификатор объекта. 

Например, создадим переменную `a`, которой присвоим значение `1`:

In [1]:
a = 1

Сначала в памяти создается объект 1 и он, очевидно, имеет свой id:

In [2]:
print(id(1)) # у вас получится другое число (все ок, не переживаем)

4306133920


Этот же идентификатор запоминается за именем переменной: 

In [3]:
print(id(a))

4306133920


Если мы создадим новую переменную, например, `b`, которая <b>ссылается</b> на 1, то и у нее будет такой же id. 

In [4]:
b = 1
print(id(b))

4306133920


In [5]:
c = a           # переменная с ссылается на a, которая в свою очередь — на 1
print(c, id(c)) 

a = 2           # перезапишем переменную а, теперь присвоим ей объект 2
print(a, id(a)) # очевидно, что значение переменной а изменится, также как и id
print(c, id(c)) # при этом это никак не повлияет на значение переменной с 

1 4306133920
2 4306133952
1 4306133920


Приведем пример чуть сложнее — пусть у нас есть некий набор чисел (в такой записи это будет называться списком) и посмотрим на идентификатор первого его элемента (тоже 1). 

In [6]:
L = [1, 3, 4]
print(id(L))
print(id(L[0])) # снова такой же id, как у 1 (фактически L[0] и есть наша 1)

4361935552
4306133920


И пара слов о том, а какие имена переменных допустимы. Предположим, я хочу назвать переменную `1qwerty` — проверим, получится или нет:

In [7]:
1qwerty = 2

SyntaxError: invalid decimal literal (271193151.py, line 1)

Выдается ошибка `SyntaxError`, которая явно свидетельствует, что что-то с названием переменной не так. Но тогда, как можно называть переменные, а как нельзя? 

Переменные в  Python подчиняются следующим правилам:

+ имя переменной может содержать только буквенно-цифровые символы и символы подчеркивания (A-Za-z, 0–9 и _)
+ имя переменной <u>должно</u> начинаться с буквы или символа подчеркивания
+ имя переменной <u>не может</u> начинаться с цифры
+ в имени переменной <u>не должно</u> быть пробелов
+ имя <u>не должно</u> совпадать со служебными (зарезервированными) словами в Python

Список зарезервированных слов можно узнать так:

In [8]:
import keyword
print(keyword.kwlist)

['False', 'None', 'True', 'and', 'as', 'assert', 'async', 'await', 'break', 'class', 'continue', 'def', 'del', 'elif', 'else', 'except', 'finally', 'for', 'from', 'global', 'if', 'import', 'in', 'is', 'lambda', 'nonlocal', 'not', 'or', 'pass', 'raise', 'return', 'try', 'while', 'with', 'yield']


+ имена переменных чувствительны к регистру
+ рекомендуется давать переменным осмысленные названия

## Изменяемые (mutable) и неизменяемые (immutable) типы данных <a name="par2"></a>

| Неизменяемые  | Изменяемые    | 
|:-------------:|:-------------:| 
| числа (`int`, `float`, `complex`)| список (`list`) | 
| логический тип (`bool`)      | словари (`dict`)      |  
| кортеж (`tuple`) |множество (`set`)       |    
| строка (`str`) |       |   
| неизменяемое множество (`frozenset`) |       |  

Да, у нас есть красивая табличка. Давайте попробуем доказать, например, что числа — действительно неизменяемый тип данных. 

In [4]:
X = 1           # пусть X ссылается на 1
Y = X           # а Y ссылается на X, то есть тоже на 1
print(X, id(X)) # собственно, тут ничего нового — 
print(Y, id(Y)) # id совпадают

X += 1          # попробуем изменить X, добавив к переменной 1
              # (аналогично можно записать X = X + 1)
print(X, id(X)) # и, действительно, значение X увеличилось на 1
print(Y, id(Y))

1 140655078730032
1 140655078730032
2 140655078730064
1 140655078730032


Вопрос — а что только что произошло? Фактически изменилось значение переменной. Но сам объект 1 никак не поменялся и изменить его мы не можем. Соответственно, все операторы (в данном примере это знак `+`), которые работают с числами, вернут нам новый объект, а не изменят уже имеющийся. 

Попробуйте аналогично разобраться, почему, например, объекты логического типа являются неизменяемыми? А строки?

Проверим и изменяемые типы, например, списки.

In [10]:
X = [1, 2, 3]    # пусть X ссылается на объект [1, 2, 3]
Y = X            # а Y ссылается на X
print(X, id(X))  # снова id совпадают
print(Y, id(Y))

X.append(4)      # изменим наш объект (на который ссылается X) и добавим к нему 4

print(X, id(X))  # очевидно, что значение объекта, на который ссылается X, изменилось,
print(Y, id(Y))  # но и значение объекта, на который ссылается Y, поменялось, 
                 # ведь метод .append() работает так, что меняет сам объект

[1, 2, 3] 4361645632
[1, 2, 3] 4361645632
[1, 2, 3, 4] 4361645632
[1, 2, 3, 4] 4361645632


Забегая чуть вперед, а что если я не хочу, чтобы значение переменной `Y` менялось? 

In [11]:
X = [1, 2, 3]   # пусть X ссылается на объект [1, 2, 3]
Y = X.copy()    # а Y сохраняет себе копию X (то есть тут буквально создается новый объект)
print(X, id(X)) # соответственно и id получаются разные
print(Y, id(Y))

X.append(4)     # изменим наш объект (на который ссылается X) и добавим к нему 4

print(X, id(X)) # объект, на который ссылается X, изменился,
print(Y, id(Y)) # но объект, на который ссылается Y, остался без изменений 

[1, 2, 3] 4361938496
[1, 2, 3] 4361732928
[1, 2, 3, 4] 4361938496
[1, 2, 3] 4361732928


Сейчас это все воспринять тяжко, но возвращайтесь к этой информации на протяжении курса и освоения нового материала.

## Типы данных — часть 1 <a name="par3"></a>

Части 2 и 3 будут в следующих конспектах. В этой же части мы посмотрим на числа и строки. 

### Числа <a name="par3.1"></a>

Числа в Python делятся на три вида: 

1. `int` — только целые числа;
2. `float` — вещественные числа / числа с плавающей точкой
3. `complex` — комплексные числа

С целыми числами все понятно, выглядит это так: 

In [12]:
print(type(1))

a = -10
print(type(a)) # применить функцию type() можно, как к самому объекту, так
               # и к переменной, которая ссылается на объект

<class 'int'>
<class 'int'>


Здесь мы обращаем внимание на `'int'`, то есть integer (целое число). Тип/класс определяет область допустимых значений объекта и набор операций над ним (например, целые числа можно складывать, умножать, делить и т.д.). 

С вещественными числами все сложнее. Тип называется `float` aka число с плавающей точкой.

In [13]:
print(type(1.2)) 

<class 'float'>


Почему точка плавает?

Посмотрим на число $2.25$. 

Его можно записать как $2.25$, как $0.225 \cdot 10$, как $22.5 \cdot 10^{-1}$, как $225 \cdot 10^{-2}$. Точка, отделяющая дробную часть от целой, будет «плавать», однако само число при этом меняться не будет, будут меняться только множители — разные степени десятки.

Также есть сложности, которые возникают при округлении чисел с плавающей точкой. Посмотрим на несколько примеров работы функции `round()`.

In [14]:
print(round(1.5)) # все ок
print(round(2.5)) # ой?
print(round(3.5)) # все ок
print(round(4.5)) # ой?

2
2
4
4


Может быть, функция всегда округляет только до четного числа?

In [15]:
print(round(1.49)) # все ок
print(round(4.51)) # все ок

1
5


Нет, все работает... Так происходит округление числа до целого, которое расположено ближе всего. Но если дробная часть равна `0.5`, то функция округляет до ближайшего четного значения.

Но это еще не все странности...

In [16]:
round(3.525, 2) # не 3.53

3.52

Эти странности связаны с тем, что число, которое мы видим (например, 3.525), не совпадает с тем, которое хранится в компьютере, потому что оно при сохранении преобразовывается и превращается из точного 3.525 в такое:

In [17]:
from decimal import Decimal
Decimal(3.525)

Decimal('3.524999999999999911182158029987476766109466552734375')

И такое число будет законно округляться до 3.52 по правилам арифметического округления. В прикладном анализе данных такие сложности редко вызывают проблемы, но знать про нее полезно, чтобы не пугаться и не удивляться неожиданным результатам. Кроме того, полезно помнить, что числа с плавающей точкой (типа `float`) не рекомендуется использовать в финансовых вычислениях и вообще в вычислениях, требующих высокой точности, поскольку они «накапливают ошибку», то есть дают неточные результаты.

С комплексными числами еще сложнее (возможно, кто-то помнит их еще со школы и тут содрогается). Используются они, в основном, в физике, квантовой механике... в общем, здесь мы сильно не будем акцентировать на них внимание, однако если вдруг встретите что-то такое `4 + 3j` — не пугайтесь.

In [18]:
print(type(4 + 3j))

<class 'complex'>


В экзамене их не будет, выдохните. 

**Со всеми типами чисел вы можете выполнять разные функции:**
* Сложение `x + 10`;
* Вычитание `x - 10`;
* Деление `x / 10`;
* Деление нацело (без остатка после запятой) `x // 10`;
* Остаток от деления на число `x % 3` (остаток от деления `x` на `3`);
* Возведение в степень `x ** 2`.

In [10]:
x = 10

print('x + 10 =', x + 10) # сложение
print('x - 10 =', x - 10) # вычитание
print('x / 10 =', x / 10) # деление
print('x // 3 =', x // 3) # деление нацело
print('x % 3 =', x % 3) # остаток от деления на число
print('x ** 2 =', x ** 2) # возведение в степень

x + 10 = 20
x - 10 = 0
x / 10 = 1.0
x // 3 = 3
x % 3 = 1
x ** 2 = 100


Аналогичные функции вы можете применять к плавающим числам.

In [11]:
x = 2.5

print('x + 10 =', x + 10) # сложение
print('x - 10 =', x - 10) # вычитание
print('x / 10 =', x / 10) # деление
print('x // 3 =', x // 3) # деление нацело
print('x % 3 =', x % 3) # остаток от деления на число
print('x ** 2 =', x ** 2) # возведение в степень

x + 10 = 12.5
x - 10 = -7.5
x / 10 = 0.25
x // 3 = 0.0
x % 3 = 2.5
x ** 2 = 6.25


**Задание №1**

Сохраните в переменную `number` число ваших полных лет.

Напечатайте на каждой отдельной строке:
1. Сколько полных лет вам будет через 10 лет;
2. Сколько полных лет вам было 5 лет назад;
3. Какое число получится, если разделить ваш возраст на 3;

In [ ]:
# YOUR CODE HERE

### Строки <a name="par3.2"></a>

Тут все проще. 

Текстовый тип данных обязательно заключается в кавычки.

Python понимает и одинарные, и двойные кавычки. Главное правило: не закрывать одинарные кавычки двойными, и наоборот. Двойные кавычки закрывают двойные, одинарные — одинарные. 

In [19]:
print(type('qwerty'))

<class 'str'>


Этот тип данных выдает нам функция `input()`, которая позволяет запрашивать данные с клавиатуры:

In [20]:
name = input()
print(name, type(name))


 <class 'str'>


Вы можете также напечатать сразу несколько строк, используя функцию `print()`

In [13]:
print('Это первая строка', ', А это вторая строка')

Это первая строка , А это вторая строка


Можете напечатать с помощью print() одновременно и строку, и значение, сохраненное в определенную переменную.

In [14]:
name = 'Анастасия'
print('В переменную "name" сохранено имя', name)

В переменную "name" сохранено имя Анастасия


Вы можете поместить также в функцию `input()` строку, которая будет показываться пользователям при запуске кода. Например:

In [17]:
name = input('Введите свое имя: ')
print('Hello, ', name, '!')

Введите свое имя: Анастасия
Hello,  Анастасия !


В примере выше вы можете заметить лишний пробел после имени и перед восклицательным знаком. По дефолту, если вы печатаете несколько переменных, разделителем между ними выступает пробел. 

Если вы хотите напечатать более сложную строку, вы можете использовать аргументы `sep` (separator):

In [22]:
print('раз', 'два', 'три') # можно не использовать аргумент sep
print('раз', 'два', 'три', sep=',') # sep = запятая
print('раз', 'два', 'три', sep=', ') # sep = запятая с пробелом

раз два три
раз,два,три
раз, два, три


И аргумент `end` (end of the line). По дефолту, это символ переноса на следующую строку `\n`.

In [28]:
name = 'Анастасия'
print('Hello,', name) # по дефолту, это символ переноса на следующую строку "\n"
print('Hello,', name, end='!') # в конце строки будет напечатан восклицательный знак

Hello, Анастасия
Hello, Анастасия!

**Задание №2**

На первой строке используйте функцию `input()` и сохраните в переменную `course` ваш любимый курс. 


На второй строке, используя функцию `print()`, напечатайте строку следующего типа: ```Мой любимый курс – это: <название курса>```

Примеры:

| `course`  | Результат    | 
|:-------------:|:-------------:| 
| `ИПУ` | `Мой любимый курс – это: ИПУ` | 
| `КПН` | `Мой любимый курс – это: КПН` |  
| `Высшая математика` | `Мой любимый курс – это: Высшая математика` |     

In [ ]:
course = #YOUR CODE HERE
# YOUR CODE HERE

### Преобразование типов <a name="par3.3"></a>

Функции для изменения типа переменных называются так же, как и сами типы. В целом, на этом можно закончить, но приведем парочку примеров.

In [21]:
n = '2300'
print(n, type(n))

n = int(n)
print(n, type(n))

2300 <class 'str'>
2300 <class 'int'>


Вот вроде бы в обоих случаях напечаталось 2300. Но в первом случае это текст, а во втором — целое число. Разница в том, какие операции нам доступны в обоих случаях. 

Предположим, что `'2300'` — это цена. Если это текст, то мы не сможем применить к нему никакие арифметические операции (проверьте сами или посмотрите следующий раздел про ошибки). Если же мы будем работать с целым числом, то сможем произвести любую арифметическую операцию (если речь все еще про цену на товар, то, например, посчитать скидку или кэшбэк). 

## Немного об ошибках <a name="par4"></a>

Здесь будут разобраны ошибки, которые чаще всего вы можете встретить. Однако, это лишь небольшая их часть. Если столкнетесь с какой-то новой ошибкой — не стесняйтесь ее загуглить :)

Помните, в начале конспекта у нас была переменная X? Попробуем ее напечатать:

In [22]:
print(x)

NameError: name 'x' is not defined

Ой, что-то пошло не так. Данна ошибка (`NameError`) буквально говорит нам, что такой переменной нет. 

Но ведь была же... 

Тут стоит вспомнить, что названия переменных чувствительны к регистру. Мы создавали переменную `X`, а напечатать пытаемся `x`, которой и правда нет.

Идем дальше. Предположим, у меня есть две переменных, которые я хочу сложить: 

In [23]:
a = 2
b = '3'

print(a + b)

TypeError: unsupported operand type(s) for +: 'int' and 'str'

Переменная `a` — это целое число, а переменная `b` — строка. Когда мы буквально пытаемся сложить число и текст, то программа ругается и не понимает, что от нее хотят. 

К слову, если поменять порядок переменных, то тип ошибки (`TypeError`) не поменяется, но будет другое пояснение. Подумайте, почему так происходит.

In [24]:
print(b + a)

TypeError: can only concatenate str (not "int") to str

И последнее на сегодня. Предположим, что у нас есть строка `'1,23'` — вроде бы десятичная дробь, но провести никакие арифметические операции мы с ней не можем. Значит, нужно попробовать эту строку преобразовать в тип `float`.

In [25]:
float('1,23')

ValueError: could not convert string to float: '1,23'

Ошибка `ValueError` возникает из-за запятой в строке — функция `float` не понимает, как можно запятую преобразовать в число.

In [26]:
float('1.23')

1.23

И теперь точно последнее. Предположим, у нас есть строка `'qwerty'` и мы хотим обратиться к последнему ее элементу — шестой букве `'y'`.

In [27]:
print('qwerty'[6])

IndexError: string index out of range

Ошибка `IndexError` показывает, что элемента с таким индексом в строке нет. Это связано с тем, что индексация в Python начинается с нуля, значит, нам здесь нужен индекс `5` (или `-1`, т.е. первый элемент с конца).

In [28]:
print('qwerty'[5])
print('qwerty'[-1])

y
y


## Разметка Markdown в Jupyter Notebook <a name="par5"></a>

В среде, в которой мы с вами работаем – в Jupyter Notebook – вы будете пользоваться в основном ячейками двух типов: **Code** и **Markdown**. В ячейке типа **Code** вы можете писать и запускать код, в ячейках **Markdown** вы можете писать и форматировать текст.

Что такое **Markdown** в целом? Это язык разметки, который можно использовать для форматирования текстовых ячеек. В Jupyter Notebook вы можете использовать разные символы для форматирования:

* *Звездочка чтобы выделить текст курсивом* - `*тут текст*`;
* **Две звездочки чтобы выделить текст жирным** - `**тут текст**`;
* Квадратные скобки и круглые скобки чтобы создать гиперссылку – `[Текст гиперссылки](ссылка)`;
* Штрихи чтобы имитировать `код` – \`тут текст или код\`

А также используйте:
* `#` для выделения заголовков первого уровня: `# Заголовок первого уровня`;
* `##` для выделения заголовков второго уровня: `## Заголовок второго уровня`;
* `###` для выделения заголовков третьего уровня: `### Заголовок третьего уровня`;

Если хотите посмотреть, как выглядит разметка в ячейке типа Markdown, дважды кликните на эту ячейку. Узнать про все возможности Markdown можно на [Datacamp](https://www.datacamp.com/tutorial/markdown-in-jupyter-notebook).

**Задача №3**

Используя ячейку типа Markdown и разметку, напечатайте в этой ячейке ниже:

1. Используя заголовок первого уровня, ваше ФИО;
2. Используя заголовок второго уровня, вашу группу;
3. Используя заголовок третьего уровня, ваше любимое животное.

## Дополнительные материалы <a name="parlast"></a>

+ Документация Python [Built-in Types](https://docs.python.org/3/library/stdtypes.html)
+ Курс на платформе Stepik «Python: основы и применение» ([ссылка на курс](https://stepik.org/course/512/syllabus))
+ Щуров И.В., Тамбовцева А.А., Жучкова С.В. —  курс «Основы программирования в Python» ([ссылка на курс](https://allatambov.github.io/pypolit/pypolit.html))
+ М. Лутц. «Изучаем Python 3», Часть II. Типы и операции.
+ Как работают числа с плавающей точкой? [YouTube](https://www.youtube.com/watch?v=U0U8Ddx4TgE)
+ Jupyter Notebook Markdown [Tutorial](https://www.datacamp.com/tutorial/markdown-in-jupyter-notebook)